### PR TITLE
Build metatensor-torch for all supported torch versions

### PR DESCRIPTION
**Pull request by Luthaf**
This tries to build metatensor_torch wheels we can publish on PyPI, following #564.

# Contributor (creator of pull-request) checklist

 - [ ] Tests updated (for new features and bugfixes)?
 - [ ] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
📚 Documentation preview 📚: https://metatensor--571.org.readthedocs.build/en/571/

### CODE REVIEW

**Comment by Luthaf** [comment]
Ok, this seems to be working! Using the wheels from the latest build, 

```
$ pip install torch
$ pip install ./metatensor_core-0.2.0.dev35-py3-none-macosx_11_0_arm64.whl
$ pip install --no-deps ./metatensor_torch-0.4.0.dev32-py3-none-macosx_11_0_arm64.whl

$ python -c "import torch; import metatensor.torch; print(torch.__version__)"
2.2.2

$ pip install torch==2.1
$ python -c "import torch; import metatensor.torch; print(torch.__version__)"
2.1.0
```

The same metatensor_torch wheel can be loaded with multiple torch versions!

----

The files sizes are not horrible either:

```
$ ls -lah metatensor_torch*
-rw-r--r--@ 1 guillaume  staff   1,9M  5 avr 14:51 metatensor_torch-0.4.0.dev32-py3-none-macosx_11_0_arm64.whl
-rw-r--r--@ 1 guillaume  staff   4,1M  5 avr 14:51 metatensor_torch-0.4.0.dev32-py3-none-macosx_11_0_x86_64.whl
-rw-r--r--@ 1 guillaume  staff   3,4M  5 avr 14:51 metatensor_torch-0.4.0.dev32-py3-none-manylinux2014_aarch64.manylinux_2_17_aarch64.whl
-rw-r--r--@ 1 guillaume  staff   3,6M  5 avr 14:51 metatensor_torch-0.4.0.dev32-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl
-rw-r--r--@ 1 guillaume  staff   2,5M  5 avr 14:51 metatensor_torch-0.4.0.dev32-py3-none-win_amd64.whl
```

----

I think the main impact is going to be on CI time. We went from ~30 CI jobs to 63 CI jobs; and building the arm64-linux wheels uses CPU emulation and is very slow (around ~20min to build the torch wheels). We could either

- not run this for all PR, only run it on release + on a schedule
- not build arm64 wheels for metatensor_torch (only metatensor_core). Users would fallback on the sdist and have to build from sources
- some other idea?

**Comment by frostedoyster** [comment]
Does this work with torch + cuda?

**Comment by Luthaf** [comment]
It should, since we are only linking to torch_cpu which should be around as well. Can you try quickly, using the right wheel from here: https://github.com/lab-cosmo/metatensor/actions/runs/8571334454/artifacts/1388709143?

**Comment by Luthaf** [comment]
It seems that the wheels still contain a reference to metatensor-core by path (`/project/python/metatensor-core`), this should be fixed.

File: .github/workflows/build-wheels.yml [hunk 1]
```diff
@@ -67,36 +67,192 @@ jobs:
...
-        if: startsWith(github.ref, 'refs/tags/metatensor-core-v')
-        uses: softprops/action-gh-release@v1
+  build-torch-wheels:
+    runs-on: ${{ matrix.os }}
+    name: ${{ matrix.name }} (torch v${{ matrix.torch-version }})
+    strategy:
+      matrix:
+        torch-version: ['1.11', '1.12', '1.13', '2.0', '2.1', '2.2']
+        arch: ['arm64', 'x86_64']
+        os: ['ubuntu-22.04', 'macos-11', 'macos-14', 'windows-2019']
```
Comment (PicoCentauri) [inline .github/workflows/build-wheels.yml:100]: Shouldn't there be another loop over python versions?

File: .github/workflows/build-wheels.yml [hunk 2]
```diff
@@ -67,36 +67,192 @@ jobs:
...
+            arch: arm64
+            rust-target: aarch64-apple-darwin
+            cibw-arch: arm64
+          - name: x86_64 Windows
+            os: windows-2019
+            arch: x86_64
+            rust-target: x86_64-pc-windows-msvc
+            cibw-arch: AMD64
+          # add the right python version for each torch version
+          - {torch-version: '1.11', python-version: '3.10', cibw-python: 'cp310-*'}
```
Comment (PicoCentauri) [inline .github/workflows/build-wheels.yml:139]: Can't you run this with Py3.12? If yes does this work? I thought we have to build the wheels for every Python major version.

**Review by PicoCentauri — COMMENTED** [review]
looks like it could work 🤣

File: .github/workflows/build-wheels.yml (hunk 1 above)
Comment (Luthaf) [reply .github/workflows/build-wheels.yml:100]: no, we only need to build the code once per torch version thanks to https://github.com/lab-cosmo/metatensor/pull/564

We just pick one python version that's compatible with the torch version for the build below.

File: .github/workflows/build-wheels.yml (hunk 2 above)
Comment (Luthaf) [reply .github/workflows/build-wheels.yml:139]: no, there are no pre-built torch 1.11 for python 3.12. The versions here are the highest Python versions for which there is a prebuilt torch wheel

File: .github/workflows/build-wheels.yml (hunk 1 above)
Comment (PicoCentauri) [reply .github/workflows/build-wheels.yml:100]: I see.

File: .github/workflows/build-wheels.yml (hunk 2 above)
Comment (PicoCentauri) [reply .github/workflows/build-wheels.yml:139]: Okay I understand.

**Review by PicoCentauri — APPROVED** [review]

**Pull request merged**